### PR TITLE
Fixing performance issue for gets.

### PR DESCRIPTION
**Pull request by sduskis**
The ChannelPool is great for scans, but causes performance impacts to Get operations.  Moving ChannelPool handling into BigtableDataGrpcClient and only using a ChannelPool for stream operations..

### CODE REVIEW

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 1]
```diff
@@ -229,18 +230,27 @@ public Row readModifyWriteRow(ReadModifyWriteRowRequest request) {
...
-    CancellationToken cancellationToken = new CancellationToken();
-    cancellationToken.addListener(new Runnable() {
-      @Override
-      public void run() {
-        readRowsCall.cancel();
-      }
-    }, executorService);
+    // Reserve a channel if this is a streaming call, since gRPC does
+
+    PooledChannel reservedChannel =  isGet ? null : channelPool.reserveChannel();
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: Channel channel = channelPool.reserveChannel();
if (isGet) {
  channel = channelPool;
}
...
CancellationToken cancellationToken = createCancellationToken(channel, readRowsCall);

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 2]
```diff
@@ -266,8 +276,24 @@ public void run() {
...
 
+  private CancellationToken createCancellationToken(final PooledChannel reservedChannel,
+      final ClientCall<ReadRowsRequest, ReadRowsResponse> readRowsCall) {
+    // If the scanner is close()d before we're done streaming, we want to cancel the RPC:
+    CancellationToken cancellationToken = new CancellationToken();
+    cancellationToken.addListener(new Runnable() {
+      @Override
+      public void run() {
+        readRowsCall.cancel();
+        if (reservedChannel != null) {
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: if (channel instanceof PooledChannel) {
  ((PooledChannel) channel).returnToPool();
}

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java (hunk 1 above)
Comment (sduskis) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: We want to use a channel from the pool without reserving it if this is a Get.  Calling  channelPool.reserveChannel() doesn't seem to make sense to me.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java (hunk 2 above)
Comment (sduskis) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: instanceof is expensive. Why can't we simply pass in a @Nullable PooledChannel?

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java (hunk 1 above)
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: Channel channel = channelPool;
if (!isGet) {
channel = channelPool.reserveChannel();
}

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java (hunk 2 above)
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: The more variables we have to keep the less readable code.
How about this 

createCancellationToken( isGet ? null : channel, readRowsCall)
or
createCancellationToken( (channel instanceof  PooledChannel)? null : channel, readRowsCall)

some upfront cost.

**Comment by sduskis** [comment]
I updated the code... how about this?

  private void returnToPool(final Channel channel) {
    if (channel != channelPool) {
      ((PooledChannel) channel).returnToPool();
    }
  }

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 3]
```diff
@@ -284,4 +299,10 @@ public void onClose(Status status, Metadata trailers) {
       }
     };
   }
+
+  private void returnToPool(final Channel channel) {
+    if (channel != channelPool) {
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: This is better, please add a comment saying that when channel is not channelPool, we are using a PooledChannel.

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 4]
```diff
@@ -229,18 +230,17 @@ public Row readModifyWriteRow(ReadModifyWriteRowRequest request) {
...
-    cancellationToken.addListener(new Runnable() {
-      @Override
-      public void run() {
-        readRowsCall.cancel();
-      }
-    }, executorService);
+    // If it's a long running scan, reserve the pool. There is a gRPC bug that makes the Channel
+    // unavailable for other streams. Reserve a channel for now. This can probably be removed once
+    // we use gRPC 0.10.0. See https://github.com/grpc/grpc-java/issues/1175 for more details
+    Channel channel = isGet ? channelPool : channelPool.reserveChannel();
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: much better. thanks!

File: bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java [hunk 5]
```diff
@@ -266,8 +266,22 @@ public void run() {
     return resultScanner;
   }
 
+  private CancellationToken createCancellationToken(final Channel channel,
+      final ClientCall<ReadRowsRequest, ReadRowsResponse> readRowsCall) {
+    // If the scanner is close()d before we're done streaming, we want to cancel the RPC:
```
Comment (kevinsi4508) [inline bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java]: close()d?

**Comment by kevinsi4508** [comment]
It looks good, please add some test to reflect the change in BigtableDataGrpcClient.java.

**Comment by sduskis** [comment]
I updated the comments and added some testing

**Pull request merged**